### DISCUSSION TITLE
Slow saves with animated gifs.

### DISCUSSION THREAD

**Discussion by sethify** [libvips/libvips#2289] [Q&A]
Hello,

I'm using the latest and greatest libvips (8.10.6). I am binding to the latest greatest GraphicsMagick for gif processing. 

I have a ridiculously large 13MB 1440x800 300 frame animated gif. I'm replacing an old system with a new system using go (our own interface) and libvips C. 

I load the file, extract the frames, resize them then recompose the image, this all happens amazingly in < 200ms. 

Saving the file on the other hand takes 77 seconds through GraphicsMagick (and worse with ImageMagick).  The solution I'm replacing spawns Gifsycile, and does the same load resize and save in about 3 seconds.

Do you have any suggestions how we could increase gif write speed?

Note: we are writing to buffers.

Thank you.

**Comment by jcupitt** [comment]
Hi @sethify,

libvips uses *magick for GIF save, and as you say, it's extremely slow. It tries a lot of very expensive optimisations: it searches every frame for sections it can reuse from previous frames, optimises the palette on each one, etc. etc. gifsicle is great, and does a much better job of gif writing (as you'd expect from a specialized tool).

We just switched GIF load libraries and we should probably look for a better GIF save library, but no one's found the time yet.

**Reply by sethify** [reply]
I've been up and down google as well, and can't find a good writer. The author of gifsicle does not recommend using it as a library (I think because he didn't want to maintain it). The gif standard is pretty set in stone, so maybe a fork of library would work?

**Reply by kleisauke** [reply]
The disadvantage of gifsicle is that it's GPL licensed. Here are some alternative GIF encoding implementations worth checking out:
https://github.com/lecram/gifenc  (~300 lines of code, public domain)
https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/gif.c (~600 lines of code, LGPLv3 license)
https://gitlab.gnome.org/Archive/byzanz/-/blob/master/gifenc/gifenc.c (~600 lines of code, based on GLib, LGPLv3 license, repo archived)
(I tried some time ago to write a `gifsave` without *magick, but I never finished it)

**Reply by sethify** [reply]
@kleisauke I just dug through those. I'm not sure any of them are fast, and none of them look supported.
@jcupitt some rust folks made some cool bindings for gifsicle: https://lib.rs/crates/gifsicle

**Reply by sethify** [reply]
@kleisauke Thank you for providing sources. Which library did you use to attempt your `gifsave`?

**Reply by jcupitt** [reply]
The gifsicle bindings look interesting, but I think making bindings ourselves is out of scope :( Plus it would be GPL, as Kleis says.

**Reply by sethify** [reply]
I agree, GPL is a can of worms that should probably remain closed.

**Reply by jcupitt** [reply]
The best solution is probably for us to use one of the simple GIF writers (`gifenc`?) and just write unoptimised GIF files. If people want better compression, they can post-process with something like gifsicle.

Have you looked at mp4 for your microscope images? It's what I generally use for complex animations: fast, small, works anywhere, simple. ffmpeg writes it well.

**Reply by sethify** [reply]
I'm not doing microscopes, I'm writing an image transform service. It will serve millions of request per hour - which is why the need for speed is there. We have part of it in testing, and it's doing well, actually we are seeing that the libvips version outperforms the magick version in almost all scenarios 👍  by a good margin.

**Reply by sethify** [reply]
@jcupitt a simple gif encoder would definitely be a good starting point.

**Reply by lovell** [reply]
Is this for Shopify? My general advice is to convert animated GIFs to either video or animated WebP, but it very much depends on which browsers are to be supported (:wave: IE 11).

https://github.com/lecram/gifenc appears to be maintained plus accepts donations so we could use some of the Open Collective budget if required. A quick look suggests it's missing transparency support plus needs to be fed a ready-built palette, which is similar to the existing quantisation logic in libvips' PNG saver (to avoid GPL https://github.com/lovell/libimagequant can be used for this).

**Reply by kleisauke** [reply]
I had a quick look at ImageMagick's source code, it looks like by default a Riemersma dither is applied during GIF encode.
https://github.com/ImageMagick/ImageMagick6/blob/94026fbfb045d27576799d84ee63320ba7cbc7aa/magick/quantize.c#L2260-L2261

From a quick test, it seems that disabling that will improve performance.
```
$ time convert dancing_banana2.lossless.webp x.gif

real    0m1.758s
user    0m1.535s
sys     0m0.457s
$ time convert dancing_banana2.lossless.webp -dither none x2.gif

real    0m0.701s
user    0m0.465s
sys     0m0.483s
$ ls -sh x*.gif
88K x2.gif  88K x.gif
```

Test image: https://storage.googleapis.com/downloads.webmproject.org/webp/images/dancing_banana2.lossless.webp

> Which library did you use to attempt your `gifsave`?

I initially based it on https://github.com/lecram/gifenc, since that one was the simplest. I'll probably have to dust off that ancient code and try to base it on the `VipsTarget` API.

**Reply by sethify** [reply]
@lovell yes, it is for Shopify. BTW. nice work on the lib! 
@kleisauke that is a good performance improvement. I'm not sure there is a way to control dithering from calls like `vips_magicksave_buffer`, unless I've completely missed the boat on saving gifs with libvips (which is always a possibility)

**Reply by sethify** [reply]
@lovell I forgot to mention but, yes, webp is an option, the webp save still took 37 seconds, which is much better, but still 10x slower than gifsicle.

**Reply by lovell** [reply]
The WebP `reduction_effort` setting should help. It's a range from 0-6 where higher means spend more CPU time to reduce file size. The default value is 4, so perhaps try reducing this as a trade off to improve throughput. It's also worth experimenting with the `min_size`, `kmin` and `kmax` settings.

https://libvips.github.io/libvips/API/current/VipsForeignSave.html#vips-webpsave
